src/pretty/lib.rs

```diff
@@ -2,7 +2,175 @@
 //! [Wadler-style](http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf)
 //! pretty-printing API.
 //!
-//! See `Doc` and
+//! Start with with the static functions of [Doc](enum.Doc.html).
+//!
+//! ## Quick start
+//!
+//! Let's pretty-print simple sexps!  We want to pretty print sexps like
+//!
+//! ```lisp
+//! (1 2 3)
+//! ```
+//! or, if the line would be too long, like
+//!
+//! ```lisp
+//! ((1)
+//!  (2 3)
+//!  (4 5 6))
+//! ```
+//!
+//! A _simple symbolic expression_ consists of a numeric _atom_ or a nested ordered _list_ of
+//! symbolic expression children.
+//!
+//! ```rust
+//! # extern crate pretty;
+//! # use pretty::*;
+//! enum SExp {
+//!     Atom(u32),
+//!     List(Vec<SExp>),
+//! }
+//! use SExp::*;
+//! # fn main() { }
+//! ```
+//!
+//! We define a simple conversion to a [Doc](enum.Doc.html).  Atoms are rendered as strings; lists
+//! are recursively rendered, with spaces between children where appropriate.  Children are
+//! [nested]() and [grouped](), allowing them to be laid out in a single line as appropriate.
+//!
+//! ```rust
+//! # extern crate pretty;
+//! # use pretty::*;
+//! # enum SExp {
+//! #     Atom(u32),
+//! #     List(Vec<SExp>),
+//! # }
+//! # use SExp::*;
+//! impl SExp {
+//!     /// Return a pretty printed format of self.
+//!     pub fn to_doc(&self) -> Doc<BoxDoc> {
+//!         match self {
+//!             &Atom(x) => Doc::as_string(x),
+//!             &List(ref xs) => {
+//!                 // `intersperse` without the itertools crate.
+//!                 let mut ds = Vec::new();
+//!                 for x in xs {
+//!                     ds.push(x.to_doc());
+//!                     ds.push(Doc::space());
+//!                 }
+//!                 if !xs.is_empty() {
+//!                     ds.pop(); // Remove trailing space.
+//!                 }
```

Comment:
Maybe write "intersperse" as this instead to avoid allocating the `Vec`.

```rust
let mut doc = Doc::nil();
for (i, x) in xs.iter().enumerate() {
    if i != 0 {
        doc = doc.append(Doc::space());
    }
    doc = doc.append(x.to_doc());
}
```